frameworks/Rust/may-minihttp/src/main.rs

```diff
@@ -159,47 +158,50 @@ impl PgConnection {
         Ok(worlds)
     }
 
-    fn updates(&self, num: usize, rand: &mut Rand32) -> Result<Vec<WorldRow>, may_postgres::Error> {
-        let mut queries = SmallVec::<[RowStream; 32]>::new();
+    fn updates(
+        &self,
+        num: usize,
+        rand: &mut Rand32,
+    ) -> Result<SmallVec<[WorldRow; 32]>, may_postgres::Error> {
+        let mut queries = SmallVec::<[_; 32]>::new();
         for _ in 0..num {
-            let random_id = rand.rand_range(1..10001) as i32;
+            let random_id = (rand.rand_u32() % 10_000 + 1) as i32;
             queries.push(
                 self.client
                     .query_raw(&self.world, utils::slice_iter(&[&random_id]))?,
             );
         }
 
-        let mut worlds = Vec::with_capacity(num);
+        let mut worlds = SmallVec::<[_; 32]>::new();
         for mut q in queries {
+            let new_random_num = (rand.rand_u32() % 10_000 + 1) as i32;
             match q.next().transpose()? {
                 Some(row) => worlds.push(WorldRow {
                     id: row.get(0),
-                    randomnumber: row.get(1),
+                    randomnumber: new_random_num,
                 }),
                 None => unreachable!(),
             }
         }
 
-        let mut update = String::with_capacity(120 + 12 * num);
-        update.push_str("UPDATE world SET randomnumber = temp.randomnumber FROM (VALUES ");
-
-        for w in &mut worlds {
-            w.randomnumber = rand.rand_range(1..10001) as i32;
-            let _ = write!(&mut update, "({}, {}),", w.id, w.randomnumber);
+        let mut params: Vec<&(dyn ToSql + Sync)> = Vec::with_capacity(num * 3);
+        for w in &worlds {
+            params.push(&w.id);
+            params.push(&w.randomnumber);
+        }
+        for w in &worlds {
+            params.push(&w.id);
         }
-        update.pop();
-        update.push_str(" ORDER BY 1) AS temp(id, randomnumber) WHERE temp.id = world.id");
 
-        self.client.simple_query(&update)?;
+        self.client.query(&self.updates[num - 1], &params)?;
         Ok(worlds)
     }
 
-    fn tell_fortune(&self) -> Result<Vec<Fortune>, may_postgres::Error> {
-        let mut items = Vec::with_capacity(80);
-        items.push(Fortune {
+    fn tell_fortune(&self) -> Result<SmallVec<[Fortune; 32]>, may_postgres::Error> {
+        let mut items: SmallVec<[_; 32]> = smallvec::smallvec![Fortune {
```

Comment:
This is going to fail with the new toolset. The array that holds fortunes needs to have dynamic sizing. We test thousands of fortunes during verification process to make sure this is the case.

Reply:
Thanks. I think SmallVec would fall back to the heap for larger allocations. The doc says " If the data exceeds this limit, the SmallVec will "spill" its data onto the heap, allocating a new buffer to hold it."